### PR TITLE
Upload streaming

### PR DESCRIPTION
**Pull request by artemredkin**
Addressing comments in the client proposal discussion, I would like to propose the following change:

`HTTPResponseDelegate.didReceivePart` now returns an optional future that could be used to indicate to the client that reads should be stopped until this future is resolved (backpressure)

`HTTPClient.Body` is now a struct with optional length and a callback for upload streaming. In addition, there are static methods to keep API the same as before.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,9 +124,11 @@ class CountingDelegate: HTTPResponseDelegate {
         // this is executed when we receive HTTP Reponse head part of the request (it contains response code and headers), called once
     }
 
-    func didReceivePart(_ buffer: ByteBuffer) {
+    func didReceivePart(_ buffer: ByteBuffer) -> EventLoopFuture<Void>? {
```
Comment (tanner0101) [inline README.md]: Returning an optional future seems unusual to me. Could we achieve the same thing by passing in an optional promise?

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 2]
```diff
@@ -267,6 +284,8 @@ internal class TaskHandler<T: HTTPClientResponseDelegate>: ChannelInboundHandler
     let redirectHandler: RedirectHandler<T.Response>?
 
     var state: State = .idle
+    var pendingRead = false
+    var mayRead: Bool = true
```
Comment (tanner0101) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Inconsistent style here

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 3]
```diff
@@ -300,33 +319,44 @@ internal class TaskHandler<T: HTTPClientResponseDelegate>: ChannelInboundHandler
...
+            }
+        }
+    }
 
-        self.state = .sent
-        self.delegate.didTransmitRequestBody(task: self.task)
+    private func writeBody(request: HTTPClient.Request, context: ChannelHandlerContext) -> EventLoopFuture<Void> {
+        if let body = request.body {
+            return body.provider { part in
+                context.writeAndFlush(self.wrapOutboundOut(HTTPClientRequestPart.body(part)))
```
Comment (tanner0101) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: You can use leading-dot syntax here to omit `HTTPClientRequestPart`

File: Sources/NIOHTTPClient/Utils.swift [hunk 4]
```diff
@@ -52,3 +53,21 @@ public class HandlingHTTPResponseDelegate<T>: HTTPClientResponseDelegate {
         throw EmptyEndHandlerError()
     }
 }
+
+class CopyingDelegate: HTTPClientResponseDelegate {
```
Comment (tanner0101) [inline Sources/NIOHTTPClient/Utils.swift]: This can probably be `final`

**Comment by weissi** [comment]
> I must be missing something... I thought `didReceivePart()` was for streaming incoming body data (e.g. GET response). Are you intending to also use this callback for streaming outgoing body data (e.g. POST)? If so, I think that is confusing to the user...

@ianpartridge no, you're not missing anything but it's important to propagate back-pressure. Let's assume we have three different nodes A, B, and C with two active HTTP connections:

    A --> B --> C

where `A --> B` is `B` download streaming a large file from `A` (eg. through a `GET`) and `B --> C` is `B` upload streaming the same file to C (eg. through a `POST`).

Now let's assume we're `B` and we prefer not to be killed by the OOM killer. We must now protect against `A` delivering us the bytes faster than `C` will accept them (or else we would need to buffer the bytes that `C` didn't accept yet in memory which would eventually lead to the OOM killer coming to find us).

How can we (as `B`) achieve this? We need to propagate the back-pressure that `C` is giving us back to `A`. For this HTTP API this would mean that we need to be able to tell the `didReceivePart` callback when we're ready to accept the next part. For blocking IO, back-pressure is easy by doing:

```swift
func didReceivePart(data: ...) {
    otherConnection.sendData(data) // blocks until `data` is fully sent
}
```

for asynchronous programming we can express the same concept as

```swift
func didReceivePart(data: ...) -> EventLoopFuture<Void> {
    return otherConnection.sendData(data) // returns a future that is fulfilled when the data has been sent
}
```

in other words: The whole thing totally works _iff_ the API allows us to indicate when we're ready for the next chunk and that's done with the change to have `didReceivePart` return a future.

And yes, this isn't actually related to upload streaming but when I was looking at the back-pressure for the upload streaming I noticed that download streaming doesn't actually support back-pressure and therefore it's in this PR :).

**Comment by ianpartridge** [comment]
OK, thanks so much for the awesome explanation.  So, if I'm understanding correctly, the new `EventLoopFuture<Void>?` return value of `didReceivePart()` means as follows:

`nil`: I've finished handling the part, you can send me another whenever you like.
An `EventLoopFuture<Void>`: I'm still handling the part, you can send me another when this future is fulfilled.

Thoughts...

Is the reason the future is optional because it avoids the cost of creating an already succeeded future in the case where the user doesn't care about propagating backpressure? Otherwise the future could be non-optional and we just ask the user to create a succeeded future.

Would returning an enum with two cases (`case done` and `case pending(EventLoopFuture<Void>)`) be clearer to the user?

**Comment by Lukasa** [comment]
I'd missed this, thanks for pointing this out @ianpartridge:

> Is the reason the future is optional because it avoids the cost of creating an already succeeded future in the case where the user doesn't care about propagating backpressure? Otherwise the future could be non-optional and we just ask the user to create a succeeded future.

I don't think we should allow a `nil` return value here. Forgetting to respect back pressure is a common source of nasty bugs that can manifest as OOMs and crashes, so the easier we can make it to propagate the backpressure the better. I think we did a good job of this on `NonBlockingFileIO.readChunked`, and I'd like to see us continue that here, so I'd be strongly in favour of requiring a future to be returned.

**Comment by weissi** [comment]
> Thoughts...
> 
> Is the reason the future is optional because it avoids the cost of creating an already succeeded future in the case where the user doesn't care about propagating backpressure? Otherwise the future could be non-optional and we just ask the user to create a succeeded future

Hah, glad you point that out! I had no idea it was optional, I don't think it should be :). @artemredkin  was that deliberate?


> Would returning an enum with two cases (`case done` and `case pending(EventLoopFuture<Void>)`) be clearer to the user?

This would be slightly cheaper at run-time (than always allocating) in the case where we don't need back-pressure (rare!) and as you point out make the API less clear. This feels like premature optimisation to me.

**Comment by artemredkin** [comment]
It was deliberate, yes. I'll change it to non-optional

**Comment by artemredkin** [comment]
Actually, reason I made it's optional is that there is no easy way to create completed future without passing in eventLoopGroup...

**Comment by Lukasa** [comment]
In general I think that's an acceptable limitation. I'm happy with the idea that any reasonable delegate implementation will need access to an event loop.

**Comment by artemredkin** [comment]
@Lukasa @weissi do you think enum may be used here, or should I pass in a non-optional promise as an argument for example?

**Comment by artemredkin** [comment]
@Lukasa I wanted delegates to have simple inits without dependencies...

**Comment by Lukasa** [comment]
Neither: I think both will be misused. If we really think it's unacceptable to require the delegate have access to an event loop then I prefer non-optional promise to the enum unless the empty case is `.iAmHappyIfSomeoneDoSesMyServerAndPromiseNotToComplain`.

**Comment by Lukasa** [comment]
> I wanted delegates to have simple inits without dependencies...

Why?

**Comment by artemredkin** [comment]
I think having something like:
```swift
let client = HTTPClient(...)
let delegate = CountingDelegate()
client.execute(request: request, delegate: delegate)
```
is slightly better than:
```swift
let client = HTTPClient(...)
let delegate = CountingDelegate(eventLoopGroup: client.eventLoopGroup)
client.execute(request: request, delegate: delegate)
```
but I see your point.

**Comment by Lukasa** [comment]
Note that you only need `EventLoop`, not `EventLoopGroup`.

**Comment by Lukasa** [comment]
Additionally, if this is a real concern, you can always pass the `EventLoop` or `Channel` to the delegate callback.

**Comment by Lukasa** [comment]
Most non-trivial delegates will need either or both of those anyway.

File: README.md (hunk 1 above)
Comment (artemredkin) [reply README.md]: See main discussion, having promise here could be error-prone

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 5]
```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
...
-
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+    typealias ChunkProvider = (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Can/should we lift this `typealias` inside the `struct Body`?

File: Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift [hunk 6]
```diff
@@ -322,4 +322,65 @@ class SwiftHTTPTests: XCTestCase {
...
+                    writer(.byteBuffer(part))
+                }
+                return httpClient.execute(request: request, delegate: delegate).future
+            } catch {
+                return httpClient.eventLoopGroup.next().makeFailedFuture(error)
+            }
+        }
+
+        let upload = try! httpClient.post(url: "http://localhost:\(httpBin.port)/post", body: body).wait()
+        let bytes = upload.body!.withUnsafeReadableBytes {
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift]: done

File: Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift [hunk 7]
```diff
@@ -322,4 +322,65 @@ class SwiftHTTPTests: XCTestCase {
...
+            buffer.writeString("1234")
+            return writer(.byteBuffer(buffer)).flatMap {
+                var buffer = allocator.buffer(capacity: 4)
+                buffer.writeString("4321")
+                return writer(.byteBuffer(buffer))
+            }
+        }
+
+        let response = try httpClient.post(url: "http://localhost:\(httpBin.port)/post", body: body).wait()
+        let bytes = response.body!.withUnsafeReadableBytes {
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift]: done

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 2 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 3 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/NIOHTTPClient/Utils.swift (hunk 4 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/Utils.swift]: done, thanks!

**Comment by ianpartridge** [comment]
If we are passing the `EventLoop` to `didReceivePart()` should we also pass it to the other callbacks?

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 5 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: good idea, fixed, thanks!

**Comment by artemredkin** [comment]
@weissi @Lukasa what do you think about passing EventLoop in didTransmitBody (and other methods)? Or do you think having it in init is better?

**Comment by Lukasa** [comment]
The advantage of passing it into those other functions is that it makes it easier to implement a stateless or semi-stateless delegate, so I might be inclined just to pass it there too. @weissi what do you think?

**Comment by weissi** [comment]
> @weissi @Lukasa what do you think about passing EventLoop in didTransmitBody (and other methods)? Or do you think having it in init is better?

I think the usual thing in delegates is to pass some object that allows you to retrieve all the information you need and also allows you to cancel it if that's available. Does the Task contain the EL and the Channel? Because if yes, then we could just pass that.

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 8]
```diff
@@ -346,7 +376,14 @@ internal class TaskHandler<T: HTTPClientResponseDelegate>: ChannelInboundHandler
                 break
             default:
                 self.state = .body
-                self.delegate.didReceivePart(task: self.task, body)
+                let future = self.delegate.didReceivePart(task: self.task, eventLoop: context.eventLoop, body)
+                self.mayRead = false
+                future.whenComplete { _ in
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: What is the future is failed? Do we still want to `read()`?

**Comment by artemredkin** [comment]
Task has an optional Channel unfortunately, but its easy to add EventLoop there

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 8 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: good catch, I think we may want to fail here, thanks!

**Comment by artemredkin** [comment]
Added EventLoop to Task

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 9]
```diff
@@ -376,12 +378,19 @@ internal class TaskHandler<T: HTTPClientResponseDelegate>: ChannelInboundHandler
...
+                future.whenComplete { result in
+                    switch result {
+                    case .success:
+                        self.mayRead = true
+                        if self.pendingRead {
+                            context.read()
+                        }
+                    case .failure(let error):
+                        self.state = .end
+                        self.delegate.didReceiveError(task: self.task, error)
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Hmm, would the user expect to receive this callback when they triggered the error themselves?

Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: What if they use `NonBlockingIO.readChunked`, for example, and just pass returned future as the return from the delegate?

Comment (ianpartridge) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: Ah good point - in that case they won't know that the read failed, so `didReceiveError()` is their only hope. Thanks :)

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 10]
```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
...
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+    struct Body {
+        typealias ChunkProvider = (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
```
[the remark this answers is not in the document]
Comment (ianpartridge) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: I wonder if the `typealias` holds its weight actually...

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 11]
```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
...
-
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+    struct Body {
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Nit: this is really a personal preference (although I think NIO style might mandate this) but I much prefer having explicit access modifiers, and just `extension` instead of `public extension`. I'm not clever enough to remember Swift's access control rules without having to think twice every time...

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 12]
```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
...
+                var buffer = ByteBufferAllocator().buffer(capacity: data.count)
+                buffer.writeBytes(data)
+                return writer(.byteBuffer(buffer))
+            }
+        }
+
+        static func string(_ string: String) -> Body {
+            return Body(length: string.utf8.count) { writer in
+                var buffer = ByteBufferAllocator().buffer(capacity: string.utf8.count)
+                buffer.writeString(string)
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Same - it seems a shame we have to copy the bytes out of the string here.

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 13]
```diff
@@ -207,13 +224,15 @@ internal extension URL {
 
 public extension HTTPClient {
     final class Task<Response> {
+        let eventLoop: EventLoop
```
Comment (ianpartridge) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: I think this needs to be `public`. If I'm correct the tests are only compiling because you have `@testable` which users won't have...

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 14]
```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
...
+        }
+
+        static func stream(length: Int? = nil, _ provider: @escaping ChunkProvider) -> Body {
+            return Body(length: length, provider: provider)
+        }
+
+        static func data(_ data: Data) -> Body {
+            return Body(length: data.count) { writer in
+                var buffer = ByteBufferAllocator().buffer(capacity: data.count)
+                buffer.writeBytes(data)
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: `HTTPPart.body` requires `IOData`, and `IOData` could be `.byteBuffer` or `.fileRegion`, so we have to write to a byte buffer. I'm not sure there is a way to avoid copying here, @weissi is this correct?

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 13 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: fixed, thanks!

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: Rewrote without typealias, seems to be fine

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 11 above)
Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: good idea, fixed, thanks!

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: @artemredkin is right, there's no way to do this.

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 11 above)
Comment (weissi) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: yes please, I think there should never be any word in front of `extension`. `public extension` for example is really bad because it means that you can add public functions without the word `public` appearing in a diff and that's an issue for all SemVer stuff.

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 12 above)
Comment (weissi) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: copies are _fast_ (unless you're copying megabytes and more). The allocation here is more troublesome and that could  be avoided.

Generally, if someone needs the maximum performance, they can always prepare a `ByteBuffer` straight away.

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 13 above)
Comment (weissi) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: yeah, if we can it might be good to remove `@testable` from the test files... But I know it's not always that straightforward. Another option might be to have to test suites: One for the public API and one that does some extra internal stuff. But this is not super high priority ofc.

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (ianpartridge) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: Shame. Could we teach `IOData` about `Data`?

File: Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift [hunk 15]
```diff
@@ -322,4 +322,65 @@ class SwiftHTTPTests: XCTestCase {
         let res = try httpClient.get(url: "https://test/ok").wait()
         XCTAssertEqual(res.status, .ok)
     }
+
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Tests/NIOHTTPClientTests/SwiftNIOHTTPTests.swift]: Added failure test and backpressure test. What do you mean by interruption? External cancellation?

**Comment by artemredkin** [comment]
Almost forgot, I also wanted to add callback for every body part upload.
@weissi @Lukasa one last question, do you think we need to return `EventLoopFuture` from `didReceiveHead` for backpressure purposes as well?

File: README.md [hunk 16]
```diff
@@ -116,26 +116,32 @@ class CountingDelegate: HTTPResponseDelegate {
 
     var count = 0
 
-    func didTransmitRequestBody() {
-        // this is executed when request is sent, called once
+    func didTransmitRequestPart(task: HTTPClient.Task<Response>, _ part: IOData) {
```
Comment (artemredkin) [inline README.md]: @weissi @ianpartridge @tanner0101 do you think we need a callback when request head is sent as well? Also, what do you think about naming pattern (transmit/sent)?

Comment (weissi) [reply README.md]: - I'd probably err on the side of 'more callbacks', @Lukasa can you think of a good use-case for a callback when the head has been sent?
- regarding naming transmit vs. send: I don't mind, think both are fine
- not sure about the `part: IOData` label, `part` is maybe a bit unclear? I would almost make it `_ part: IOData`, don't think the label adds a lot of value there given the type `IOData`? But this is really not super important :)

**Comment by Lukasa** [comment]
Yeah, more callbacks is always better, go for it.

File: README.md (hunk 16 above)
Comment (artemredkin) [reply README.md]: but it's already `_ part: IOData` :)

Comment (weissi) [reply README.md]: whoops :), sorry

**Comment by artemredkin** [comment]
@weissi are you happy with the backpressure/failure tests?
@Lukasa do you think there is any value in returning Future from `didReceiveHead` for backpressure (maybe like opening file for writing is return code is 2xx or something)?

**Comment by Lukasa** [comment]
I'm in general in favour of being able to exert back pressure at as many places as possible, yeah.

**Comment by artemredkin** [comment]
@weissi @Lukasa @ianpartridge @tanner0101 do you have any additional comments on this PR?

File: Sources/NIOHTTPClient/HTTPHandler.swift [hunk 17]
```diff
@@ -18,25 +18,39 @@ import NIOConcurrencyHelpers
...
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+extension HTTPClient {
+    public struct Body {
+        public var length: Int?
+        public var provider: (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
```
Comment (tanner0101) [inline Sources/NIOHTTPClient/HTTPHandler.swift]: Using a struct here would give us more flexibility to add / deprecate methods in the future. It would also allow users of the package to add conformances / extensions. 

Something like `Body.StreamWriter` or `Body.Provider`:

```swift
extension HTTPClient.Body 
    var writer: @escaping (StreamWriter) -> EventLoopFuture<Void>

    public struct StreamWriter {
        let closure: @escaping (IOData) -> EventLoopFuture<Void>
        public func write(_ data: IOData) -> EventLoopFuture<Void> {
            return self.closure(data)
        }
    } 

    public static func stream(length: Int? = nil, _ writer: (StreamWriter) -> EventLoopFuture<Void>) -> 
       ...
    }
}
...
```

This also reduces the mental load to understand the code a bit. Here's example usage:

```swift
let body: HTTPClient.Body = .stream(length: 50) { stream in
    var request = try! Request(url: "http://localhost:\(httpBin.port)/events/10/1")
    request.headers.add(name: "Accept", value: "text/event-stream")

    let delegate = CopyingDelegate { part in
        stream.write(.byteBuffer(part))
    }
    return httpClient.execute(request: request, delegate: delegate).future
}
```

I'd also make the `Body.provider` property private if possible

Comment (artemredkin) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: good idea, refactored, thanks!

File: Sources/NIOHTTPClient/HTTPHandler.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOHTTPClient/HTTPHandler.swift]: We could, but it won't help much.

Our underlying writing mechanism heavily uses the reader/writer indices of `ByteBuffer`. This means we need a `ByteBuffer` *eventually*, which means we need to copy *eventually*. At least in this case we can do it as close to the user as possible, making it clearer in profiling, rather than burying that copy deep in NIO's I/O loop.
